### PR TITLE
Stochastic Weighted Average (SWA) support in PyTorch 1.6

### LINKED ISSUE #793: Stochastic Weighted Average (SWA) support in torch 1.6
## 🚀 Feature
SWA for YOLOv5. This would be an experiment to determine if SWA offers improvements over the current EMA implementation.
https://pytorch.org/blog/pytorch-1.6-now-includes-stochastic-weight-averaging/

Tests would be YOLOv5l COCO training from scratch to 300 epochs with 1) no averaging, 2) EMA, 3) SWA.

It should be noted that the current cosine LR scheduler provides similar functionality to EMA and SWA, but in the last YOLOv3 tests we ran when implementing EMA we still saw improvements over the baseline 'no averaging' scenario even with the cosine LR scheduler in place. See https://github.com/rwightman/pytorch-image-models/issues/102#issuecomment-601424476

### PR DESCRIPTION
**Pull request by glenn-jocher**
This PR implements some experimental SWA functionality to help determine if SWA offers improvements over the current EMA implementation.
https://pytorch.org/blog/pytorch-1.6-now-includes-stochastic-weight-averaging/

Tests would be YOLOv5l COCO training from scratch to 300 epochs with 1) no averaging, 2) EMA, 3) SWA.

It should be noted that the current cosine LR scheduler provides similar functionality to EMA and SWA, but in the last YOLOv3 tests we ran when implementing EMA we still saw improvements over the baseline 'no averaging' scenario even with the cosine LR scheduler in place. See https://github.com/rwightman/pytorch-image-models/issues/102#issuecomment-601424476

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhanced YOLOv5 data handling, introduced SWA for BatchNorm, and improved detection tracking.

### 📊 Key Changes
- 📈 Added automatic conversion of input images from `uint8` to `float32` and normalized them from 0-255 to 0.0-1.0 range in `models/yolo.py`.
- 🔄 Changed `info()` method in `yolo.py` to accept a `verbose` argument, allowing for more detailed model information output.
- 👀 Switched to use a `set` for tracking detected objects in `test.py` for more efficient detection handling.
- ✨ Integrated Stochastic Weight Averaging (SWA) BatchNorm update in `train.py` during the final epoch.
- 🛠 Simplified model retrieval from EMA (Exponential Moving Average) without needing to check for the 'module' attribute in `train.py`.

### 🎯 Purpose & Impact
- 🖼 The automatic image type conversion aims to reduce preprocessing steps and potential errors when feeding images into the model.
- 🕵️‍♂️ The detection improvements in `test.py` aim to enhance the model's object tracking accuracy and efficiency during testing.
- 🎭 Introducing SWA BatchNorm update is expected to increase the robustness and generalization of the model by averaging the model weights over the last few epochs, potentially leading to better performance on a variety of datasets.
- 🚀 The simplifications in model handling streamline the process, making the codebase easier to maintain and understand, which can positively influence both development time and the user experience.

These changes can significantly enhance the model's usability and performance, offering an upgraded experience for developers and end-users of the YOLOv5 object detection system. 🚅💡

### CODE REVIEW

**Comment by glenn-jocher** [comment]
Experiments with running swa batchnorm computation on val set before computing mAP. Results show decrease in mAP in all cases unfortunately. This does not implement actual SWA but rather just the final step of BN computation.
```bash
python train.py --data coco.yaml --weights yolov5l.pt
```

Baseline
```bash
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.477
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.664
```

model.train() mode instead of model.eval() mode
```bash
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.464
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.651
```

val2017 swa bn  `path=val_path, hyp=None, augment=False, cache=False, pad=0.5, rect=True`
```bash
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.474
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.661
```

val2017 swa bn  `path=val_path, hyp=None, augment=False, cache=False, pad=0.5, rect=False`
```bash
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.472
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.660
```

EDIT: my quick conclusion is that SWA is not going to offer any anything beyond what EMA is already providing, but further work to actually implement the entire SWA method would be required. No time currently, but I'll leave this open for future work.

**Comment by pfeatherstone** [comment]
@glenn-jocher why did you run swa on batchnorm layers only, not all the weights?

**Comment by glenn-jocher** [comment]
@pfeatherstone so far this is only an experiment to see the effect of the final SWA step, which is a onetime BN update after training. This single change produced poor results.
